R/assign_datetime.R

```diff
@@ -0,0 +1,167 @@
+#'     10L,     "MD1",       378,             "28-Jan-20",   "1-Feb-20",    NA,
+#'     11L,     "MD1",       378,             "12-Feb-20",   "18-Feb-20",   NA,
+#'     12L,     "MD1",       379,             "10-UNK-2020", "20-UNK-2020", NA,
+#'     13L,     "MD1",       379,             NA,            NA,            NA,
+#'     14L,     "MD1",       379,             NA,            "17-Feb-20",   NA
+#'   )
+#'
+#' cm1 <-
+#'   assign_datetime(
+#'     raw_dat = md1,
+#'     raw_var = "MDBDR",
+#'     raw_fmt = "d-m-y",
+#'     raw_unk = c("UN", "UNK"),
+#'     tgt_var = "CMSTDTC"
+#'   )
+#'
+#' cm1
+#' problems(cm1$CMSTDTC)
+#'
+#' cm_inter <-
+#'   tibble::tibble(
+#'     oak_id = 1L:14L,
+#'     raw_source = "MD1",
+#'     patient_number = c(
+#'       375, 375, 376, 377, 377, 377, 377, 378,
+#'       378, 378, 378, 379, 379, 379
+#'     ),
+#'     CMTRT = c(
+#'       "BABY ASPIRIN",
+#'       "CORTISPORIN",
+#'       "ASPIRIN",
+#'       "DIPHENHYDRAMINE HCL",
+#'       "PARCETEMOL",
+#'       "VOMIKIND",
+#'       "ZENFLOX OZ",
+#'       "AMITRYPTYLINE",
+#'       "BENADRYL",
+#'       "DIPHENHYDRAMINE HYDROCHLORIDE",
+#'       "TETRACYCLINE",
+#'       "BENADRYL",
+#'       "SOMINEX",
+#'       "ZQUILL"
+#'     ),
+#'     CMINDC = c(
+#'       "NA",
+#'       "NAUSEA",
+#'       "ANEMIA",
+#'       "NAUSEA",
+#'       "PYREXIA",
+#'       "VOMITINGS",
+#'       "DIARHHEA",
+#'       "COLD",
+#'       "FEVER",
+#'       "LEG PAIN",
+#'       "FEVER",
+#'       "COLD",
+#'       "COLD",
+#'       "PAIN"
+#'     )
+#'   )
+#'
+#' cm2 <-
+#'   assign_datetime(
+#'     raw_dat = md1,
+#'     raw_var = "MDBDR",
+#'     raw_fmt = "d-m-y",
+#'     tgt_var = "CMSTDTC",
+#'     tgt_dat = cm_inter
+#'   )
+#'
+#' cm2
+#' problems(cm2$CMSTDTC)
+#'
+#' @export
+assign_datetime <-
+  function(raw_dat,
+           raw_var,
+           raw_fmt,
+           tgt_var,
+           raw_unk = c("UN", "UNK"),
+           tgt_dat = NULL,
+           id_vars = oak_id_vars(),
+           .warn = TRUE) {
+    admiraldev::assert_character_vector(raw_var)
+    admiraldev::assert_character_scalar(tgt_var)
+    admiraldev::assert_character_vector(id_vars)
+    assertthat::assert_that(contains_oak_id_vars(id_vars),
+      msg = "`id_vars` must include the oak id vars."
+    )
+    admiraldev::assert_data_frame(raw_dat, required_vars = rlang::syms(c(id_vars, raw_var)))
+    admiraldev::assert_data_frame(tgt_dat, required_vars = rlang::syms(id_vars), optional = TRUE)
+    admiraldev::assert_character_vector(raw_unk)
+    admiraldev::assert_logical_scalar(.warn)
+
+    tgt_val <-
+      create_iso8601(!!!raw_dat[raw_var],
+        .format = raw_fmt,
+        .na = raw_unk,
+        .warn = .warn
+      )
+
+    der_dat <-
+      raw_dat |>
+      dplyr::select(c(id_vars, raw_var)) |>
+      dplyr::mutate("{tgt_var}" := tgt_val) |> # nolint object_name_linter()
+      dplyr::select(-rlang::sym(raw_var))
```

Comment:
This works well with just the date as in the example. We need to update it to make sure it works with the date and time when the input is in two different variables.

```R
cm1 <-
   assign_datetime(
     raw_dat = md1,
     raw_var = c("MDEDR","MDETM"),
     raw_fmt = c("d-m-y", "H:M:S"),
     raw_unk = c("UN", "UNK"),
     tgt_var = "CMSTDTC"
   )
```
Results in error.

This is similar to `create_iso8601(md1$MDEDR, md1$MDETM, .format = c("d-m-y", "H:M:S"), .check_format = FALSE)`

Rest everything looks good to me. Please add a test case as well.

Reply:
Thanks @rammprasad ! It's a silly mistake, `dplyr::select(-rlang::sym(raw_var))` should have been simply `dplyr::select(-raw_var)`. I will fix, and add a test case.

Reply:
@rammprasad: fixed now.